File: 2577/Solution.go
func minimumTime(grid [][]int) int {
    if grid[0][1] > 1 && grid[1][0] > 1 {
        return -1
    }

    R, C := len(grid), len(grid[0])
    q := &Pq{}
    heap.Init(q)
    heap.Push(q, Cell{0, 0, 0})
    dirs := [][]int{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}

    seen := make([][]bool, R)
    for i := range seen {
        seen[i] = make([]bool, C)
    }
    seen[0][0] = true

    for q.Len() > 0 {
        cur := heap.Pop(q).(Cell)
        if cur.x == R - 1 && cur.y == C - 1 {
            return cur.t
        }
        for _, d := range dirs {
            x := cur.x + d[0]
            y := cur.y + d[1]

            if x >= 0 && y >= 0 && x < R && y < C && !seen[x][y] {
                seen[x][y] = true
                if cur.t + 1 < grid[x][y] {
                    z := grid[x][y] - cur.t - 1
                    if z % 2 != 0 {
                        z++
                    }
                    heap.Push(q, Cell{x, y, cur.t + z + 1})
                } else{
                    heap.Push(q, Cell{x, y, cur.t + 1})
                }
            }
        }
    }
    return -2
}

type Cell struct {
    x, y, t int
}

type Pq []Cell

func (q Pq) Len() int { return len(q) }
func (q Pq) Less(i, j int) bool { return q[i].t < q[j].t }
func (q Pq) Swap(i, j int) { q[i], q[j] = q[j], q[i] }
func (q *Pq) Push(x interface{}) { *q = append(*q, x.(Cell)) }
func (q *Pq) Pop() interface{} { 
    old := *q
    n := len(old)
    item := old[n-1]
    *q = old[0 : n-1]
    return item
}

File: 2577/Solution.java
class Solution {
    public int minimumTime(int[][] grid) {
        if(grid[0][1] > 1 && grid[1][0] > 1) return -1;

        int R = grid.length, C = grid[0].length;
        PriorityQueue<int[]> q = new PriorityQueue<>((a, b) -> a[2] - b[2]);
        int[][] dir = new int[][]{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        boolean[][] v = new boolean[R][C];

        q.add(new int[]{0, 0, 0});
        v[0][0] = true;

        while(!q.isEmpty()){
            int[] cur = q.poll();

            if(cur[0] == R - 1 && cur[1] == C - 1) return cur[2];

            for(int[] d : dir){
                int x = cur[0] - d[0];
                int y = cur[1] - d[1];
                if(x >= 0 && y >= 0 && x < R && y < C && !v[x][y]){
                    v[x][y] = true;
                    if(cur[2] + 1 < grid[x][y]){
                        int z = grid[x][y] - cur[2] - 1;
                        if(z % 2 != 0) z++;
                        q.add(new int[]{x, y, cur[2] + z + 1});
                    }else{
                        q.add(new int[]{x, y, cur[2] + 1});
                    }
                }
            }
        }
        
        return -2;
    }
}
